File: Data Structures & Algorithms/combination-target-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        const res = [];
        nums.sort((a, b) => a - b);

        const backtrack = (combination, remaining, start) => {
            if (remaining === 0) {
                res.push([...combination]);
                return;
            }

            if (remaining < 0) {
                return;
            }

            for (let i = start; i < nums.length; i++) {
                combination.push(nums[i]);
                backtrack(combination, remaining - nums[i], i);
                combination.pop();
            }
        };

        backtrack([], target, 0);
        return res;
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        results = []
        candidates.sort()

        def backtrack(start_index, current_sum, current_combination):
            if current_sum == target:
                results.append(list(current_combination))
                return
            if current_sum > target:
                return
            
            for i in range(start_index, len(candidates)):
                if i > start_index and candidates[i] == candidates[i - 1]:
                    continue
                
                current_combination.append(candidates[i])
                backtrack(i + 1, current_sum + candidates[i], current_combination)
                current_combination.pop()
        
        backtrack(0, 0, [])
        return results